### PR TITLE
Add Karn Keyboard

### PR DESCRIPTION
**Pull request by robcmills**
## Description

Added a new folder `keyboards/karn`, with the necessary files to build the firmware for the Karn keyboard.

https://github.com/robcmills/karn-keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by robcmills** [comment]
This keyboard and PR is based on a very similar keyboard (the Cantor) that has already been added: https://github.com/qmk/qmk_firmware/pull/16552

File: keyboards/karn/config.h
Comment (waffle87) [inline keyboards/karn/config.h:10]:
```suggestion
```

File: keyboards/karn/rules.mk
Comment (waffle87) [inline keyboards/karn/rules.mk:3]: This can't live at the keyboard level, as it will cause compilation errors when using the QMK Configurator or any other keymap that doesn't have the necessary tap dance code.

File: keyboards/karn/readme.md
Comment (waffle87) [inline keyboards/karn/readme.md:16]: These images should be rehosted on Imgur.

File: keyboards/karn/readme.md
Comment (waffle87) [inline keyboards/karn/readme.md:16]: These images should be rehosted on Imgur.

File: keyboards/karn/keymaps/colemak/keymap.c
Comment (waffle87) [inline keyboards/karn/keymaps/colemak/keymap.c:21]:
```suggestion
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/karn/info.json
Comment (zvecr) [inline keyboards/karn/info.json:5]:
```suggestion
```

File: keyboards/karn/info.json
Comment (zvecr) [inline keyboards/karn/info.json:15]:
```suggestion
    "development_board": "blackpill_f401",
```

File: keyboards/karn/config.h
Comment (zvecr) [inline keyboards/karn/config.h:40]: Shouldnt be enabled at the keyboard level, and should be moved to keymaps where required.
```suggestion
```

File: keyboards/karn/readme.md
Comment (zvecr) [inline keyboards/karn/readme.md:75]: Shouldnt be here, and moved to a `keymaps/default/readme.md`

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/karn/readme.md
Comment (robcmills) [reply keyboards/karn/readme.md]: [karn 90dcbd1ff6] Host images on imgur

File: keyboards/karn/keymaps/colemak/keymap.c
Comment (robcmills) [reply keyboards/karn/keymaps/colemak/keymap.c]: [karn 1b888a811b] Remove #define left in colemak/keymap.c

File: keyboards/karn/info.json
Comment (robcmills) [reply keyboards/karn/info.json]: @zvecr Removing this causes the build to fail: https://github.com/qmk/qmk_firmware/runs/11340797021
Where should the bootloader be specified?

File: keyboards/karn/rules.mk
Comment (robcmills) [reply keyboards/karn/rules.mk:3]: @waffle87 Ok thanks for letting me know. Where should I move it to? (trying to figure this out from the docs but having a hard time)

File: keyboards/karn/config.h
Comment (robcmills) [reply keyboards/karn/config.h]: @zvecr Ok, not sure if this is correct, but moved these to `keyboards/karn/keymaps/colemak/config.h` (which is my customized keymap). And I cleaned up the default keymap to not include any of these defines or tap dance.

[karn 88d9db02bb]

File: keyboards/karn/rules.mk
Comment (robcmills) [reply keyboards/karn/rules.mk]: Trying moving just the `TAP_DANCE_ENABLE` to `keyboards/karn/keymaps/colemak/rules.mk`

File: keyboards/karn/readme.md
Comment (robcmills) [reply keyboards/karn/readme.md:75]: [karn 1afcf5e783] Move keymap ascii diagram from root readme to keymaps/colemak/readme

File: keyboards/karn/info.json
Comment (robcmills) [reply keyboards/karn/info.json]: Welp, looks like the subsequent commit to change the dev board fixed it? https://github.com/qmk/qmk_firmware/pull/19846/commits/a37bb6ff00516027e5d03303b66e730785b325dc
Anyways, looks like builds are passing now 👍

**Review by waffle87 — APPROVED** [review]

**Comment by robcmills** [comment]
@zvecr Thank you for reviewing this. I think I have addressed all your review comments. I resolved the comment threads that seemed straightforward, and left the others open for your review. Please let me know if there are any other changes needed. 🙏

File: keyboards/karn/readme.md
Comment (drashna) [inline keyboards/karn/readme.md:16]: This uses a smaller, but high res version of the images, that is more suited to github's markdown
```suggestion
![karn-2-left](https://i.imgur.com/vm6XFyIh.jpeg)

![karn-2-full](https://i.imgur.com/R7WnlC3h.jpeg)

![karn-2-pcb](https://i.imgur.com/EQqyyEDh.png)
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by robcmills** [comment]
@zvecr @drashna Thanks for the reviews. I am experiencing an odd bug with Github whereby if I request review from one of you, it removes the review request from the other automatically. I can't seem to get it into a state where review is requested from both. 
Anyway, I think this PR is ready for another (final?) review.

File: keyboards/karn/keymaps/colemak/config.h
Comment (drashna) [inline keyboards/karn/keymaps/colemak/config.h:4]:
```suggestion
#pragma once

// Equivalent to zmk behavior-hold-tap tap-preferred flavor
```

File: keyboards/karn/info.json
Comment (waffle87) [inline keyboards/karn/info.json:19]:
```suggestion
    },
    "bootmagic": {
        "matrix": [0, 1]
    },
    "split": {
        "enabled": true,
        "bootmagic": {
            "matrix": [4, 4]
        }
    },
```

File: keyboards/karn/config.h
Comment (waffle87) [inline keyboards/karn/config.h:26]:
```suggestion
```

File: keyboards/karn/rules.mk
Comment (waffle87) [inline keyboards/karn/rules.mk:1]:
```suggestion
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/karn/info.json
Comment (robcmills) [reply keyboards/karn/info.json:19]: So it looks like there is a general move from the `config.h` to conveying everything in `info.json`. Is that an accurate description of the current state of qmk?

**Comment by robcmills** [comment]
@waffle87 Looks like some of the changes you suggested caused the build to fail:

https://github.com/qmk/qmk_firmware/pull/19846/checks?check_run_id=11439752925

```
Ψ Compiling keymap with gmake --jobs=41 --output-sync=target karn:default


QMK Firmware 0.19.12
☒ Not including data from file: keyboards/karn/info.json
☒ 	split: Additional properties are not allowed ('bootmagic' was unexpected)
☒ karn: No LAYOUTs defined! Need at least one layout defined in info.json.
Making karn with keymap default

☒ Not including data from file: keyboards/karn/info.json
☒ 	split: Additional properties are not allowed ('bootmagic' was unexpected)
☒ karn: No LAYOUTs defined! Need at least one layout defined in info.json.

 * No bootloader specified. Please set an appropriate bootloader in your rules.mk or info.json.        [ERRORS]

platforms/avr/bootloader.mk:142: *** Invalid BOOTLOADER.  Stop.
gmake: *** [Makefile:414: karn:default] Error 1
```

I will try to fix...

**Comment by waffle87** [comment]
Ah my bad — looks like `split.bootmagic` key only exists on `develop` currently.

**Review by waffle87 — APPROVED** [review]

**Comment by robcmills** [comment]
@zvecr @drashna No rush, but still waiting on approval of this PR. Just pinging before it gets too stale. Thanks 🙏

**Review by drashna — APPROVED** [review]

**Pull request merged**
